### DISCUSSION TITLE
API via JSON request

### DISCUSSION THREAD

**Discussion by Cb5782** [iopool/community#5] [Ideas]
Hello to all,

Does anyone know whether a JSON query is possible to the IO_ Pool API?

Did anymore have already tried to query IO_Pool API, from a GCE Electronic product (IPX800_V4 or IPX800_V5)?

On the top of it, IPX800_V5 is MQTT certified.

Do you know if it works?

It could be useful to have this functionality (in 1 URL, you have 1 http query including all parameters; no need to develop JavaScript)

Regards,
Christophe

**Comment by mguyard** [comment]
Hi,

iopool API already return JSON as explain here : https://help.iopool.com/fr/articles/5537423-api-publique-iopool
In doc example: 

![JSON](https://downloads.intercomcdn.com/i/o/382869251/0a2e0724a2940bbe0ef21119/carbon+%284%29.png)

It's a json. You only need to parse json content and use what you need

**Comment by Cb5782** [comment]
Yes, but I need to send this HTTP GET Request from my "Domotique" system at home, and that's why I included in the emails yesterday the responsible of the dev. & software of GCE electronics, in the loop.
GCE is indeed performing some validation tests to put in production an update the soft of their latest product on the market : the IPX800_V5.This will allow to give the possibility for the users to have headers in the PUSH they can do (the query are called PUSH).
Once again, thank you Marc :-)I appreciated our exchanges.
Christophe
---

    Le jeudi 24 mars 2022, 15:53:18 UTC+1, Marc GUYARD ***@***.***> a écrit :  
 
 


Hi,

iopool API already return JSON as explain here : https://help.iopool.com/fr/articles/5537423-api-publique-iopool
In doc example:



It's a json. You only need to parse json content and use what you need

**Comment by Cb5782** [comment]
Hi Marc,
Just for my information, why did you decide to create a JSON "imbricated" response from your server?Is it s standard, or what I understand it is because it is specifically appropriated to JEEDOM domotique management software for the queries and parsing of the JSON answers.
I am asking you the question, because to improve the situation on GCE side, I have created the following topic:https://forum.gce-electronics.com/t/comment-faire-une-requete-http-en-get-avec-header-sur-une-v5/15115
... and unfortunately, it seems to be a second problem.
Even if GCE release a new version of the V5 software management, that will allow to perform HTTP query in GET (with headers), the parsing system will NOT be able to manage the parsing of the JSON answer from IO_Pool server, because normally, the JSON response should normally be "NOT imbricated".(each returned variable/values shall be at the first level; not some values imbricated inside a first level returned value)
I had a look to standard practice and it seems that all connected objects that exchange data via this HTTP query and JSON answer protocol, are with NOT IMBRICATED JSON answers...
That's why I am wondering why this specific choice of imbricated JSON parameters/values, in the returned answer from the server?
Thanks for your precious feed-backs,Christophe
---
    Le jeudi 24 mars 2022, 15:53:18 UTC+1, Marc GUYARD ***@***.***> a écrit :  
 
 


Hi,

iopool API already return JSON as explain here : https://help.iopool.com/fr/articles/5537423-api-publique-iopool
In doc example:



It's a json. You only need to parse json content and use what you need

**Comment by mguyard** [comment]
Hi,

> why did you decide to create a JSON "imbricated" response from your server

It's not my decision. This is the response from iopool API.

> I am asking you the question, because to improve the situation on GCE side, I have created the following topic:https://forum.gce-electronics.com/t/comment-faire-une-requete-http-en-get-avec-header-sur-une-v5/15115... and unfortunately, it seems to be a second problem.

The issue is not the imbricated format. JSON is only a Javascript object format. So it's allow and recommanded (to improve reading) to use imbricated content.

> Even if GCE release a new version of the V5 software management, that will allow to perform HTTP query in GET (with headers), the parsing system will NOT be able to manage the parsing of the JSON answer from IO_Pool server, because normally, the JSON response should normally be "NOT imbricated".(each returned variable/values shall be at the first level; not some values imbricated inside a first level returned value)

Most of mature home automation solution is able to parse JSON to permit access to subpart of response.

-----

One solution for you, as GCE don't support header addition and JSON parsing in multilevel, the best way for you is (choose your better method) :

1. Create a script on one of your server (in language you want - excluding JS who is a client language, in this case you need a server language), able to receive a query callable by GCE, who reply a non imbricated content (so you need to rewrite iopool response to a json parsable by GCE)
2. Install a more complete domotic solution like Jeedom (also able to interract with IPX 2 to 4 https://doc.jeedom.com/fr_FR/plugins/automation%20protocol/ipx800v4/, IPX5 is in beta - main developper also use IPX) and use my plugin for iopool interractions.

Jeedom is free (some plugin can have a cost) and can be installed on a RPI so it's not an expensive solution and don't need to change most of your usage.

**Reply by mguyard** [reply]
For the first purpose, this script can works :

```
<?php

// Request iopool API
$curl = curl_init();
curl_setopt_array($curl, array(
  CURLOPT_URL => 'https://api.iopool.com/v1/pool/<yourPoolId>',
  CURLOPT_RETURNTRANSFER => true,
  CURLOPT_ENCODING => '',
  CURLOPT_MAXREDIRS => 10,
  CURLOPT_TIMEOUT => 0,
  CURLOPT_FOLLOWLOCATION => true,
  CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_1_1,
  CURLOPT_CUSTOMREQUEST => 'GET',
  CURLOPT_HTTPHEADER => array(
    'x-api-key: <yourApiKey>'
  ),
));
$response = curl_exec($curl);
$info = curl_getinfo($curl);
curl_close($curl);

// Parse JSON
$jsonDecoded = json_decode($response, true);
//echo var_export($jsonDecoded);
//echo var_export($info);

// If HTTP code isn't 200 - Invalid request
if ($info['http_code'] == 200) {

  // Flatten JSON
  $flatten = new RecursiveIteratorIterator(new RecursiveArrayIterator($jsonDecoded));
  $flattenOutput = array();
  foreach($flatten as $key => $value) {
    $flattenOutput[$key] = $value;
  }

  // Return a flatten JSON
  echo json_encode($flattenOutput, JSON_PRETTY_PRINT|JSON_UNESCAPED_UNICODE);
  
} else {
  // Return error Message
  echo json_encode(array('code' => $info['http_code'], 'response' => $jsonDecoded), JSON_PRETTY_PRINT|JSON_UNESCAPED_UNICODE);
}
```

It will return something like this :

> {
>     "id": "xxxxxx-xxx-xxxxx-xxxx-xxxxxxx",
>     "title": "Pool",
>     "temperature": 21.568278169673192,
>     "ph": 8.379251700680273,
>     "orp": 290,
>     "mode": "STANDARD",
>     "isValid": true,
>     "ecoId": "XXXXXXXXX",
>     "measuredAt": "2022-03-28T12:19:00.000Z",
>     "hasAnActionRequired": true,
>     "filtrationDuration": 7
> }

**Comment by Cb5782** [comment]
Example below of a standard JSON answer ... (see picture and syntax)


---
    Le lundi 28 mars 2022, 11:59:38 UTC+2, christophe boyer ***@***.***> a écrit :  
 
  Hi Marc,
Just for my information, why did you decide to create a JSON "imbricated" response from your server?Is it s standard, or what I understand it is because it is specifically appropriated to JEEDOM domotique management software for the queries and parsing of the JSON answers.
I am asking you the question, because to improve the situation on GCE side, I have created the following topic:https://forum.gce-electronics.com/t/comment-faire-une-requete-http-en-get-avec-header-sur-une-v5/15115
... and unfortunately, it seems to be a second problem.
Even if GCE release a new version of the V5 software management, that will allow to perform HTTP query in GET (with headers), the parsing system will NOT be able to manage the parsing of the JSON answer from IO_Pool server, because normally, the JSON response should normally be "NOT imbricated".(each returned variable/values shall be at the first level; not some values imbricated inside a first level returned value)
I had a look to standard practice and it seems that all connected objects that exchange data via this HTTP query and JSON answer protocol, are with NOT IMBRICATED JSON answers...
That's why I am wondering why this specific choice of imbricated JSON parameters/values, in the returned answer from the server?
Thanks for your precious feed-backs,Christophe
---
    Le jeudi 24 mars 2022, 15:53:18 UTC+1, Marc GUYARD ***@***.***> a écrit :  
 
 


Hi,

iopool API already return JSON as explain here : https://help.iopool.com/fr/articles/5537423-api-publique-iopool
In doc example:



It's a json. You only need to parse json content and use what you need
